WIP: Refactor Axis component removing external dependencies

Kane610 opened the pull request:
## Description:

Rewrite axis to asyncio and get rid of external dependencies such as gstreamer to improve portability.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3673

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

Kane610 committed:
Async rewrite

Kane610 committed:
Async rewrite

Kane610 committed:
Device and events now working with async core

Kane610 committed:
Device and events now working with async core

Kane610 committed:
More async than before

Kane610 committed:
More async than before

Kane610 committed:
Methods have moved

Kane610 committed:
Methods have moved

Kane610 committed:
Remove check if parameter serial number is available since library handles reconnection anyway and we can expect user to set up configuration correctly

Kane610 committed:
Remove check if parameter serial number is available since library handles reconnection anyway and we can expect user to set up configuration correctly

Kane610 committed:
Serial number is the only unique identifier, checks are still needed

Kane610 committed:
Merge branch 'async_rewrite_axis' of github.com:Kane610/home-assistant into async_rewrite_axis

Conflicts:
	homeassistant/components/axis.py

Kane610 committed:
No async for setup_device

Kane610 committed:
Bump axis to 13 in preparation of friday the 13th

Kane610 commented:
I'd be happy if someone would do a quick review of https://github.com/Kane610/axis as well

andrey-git reviewed (commented):
General comment: Please add yourself to https://github.com/home-assistant/home-assistant/blob/dev/CODEOWNERS

balloob commented on homeassistant/components/axis.py:
```diff
@@ -143,16 +148,18 @@ def configuration_callback(callback_data):
-            _LOGGER.info("Stopping metadatastream for %s.", serialnumber)
-            device.stop_metadatastream()
+            _LOGGER.info("Stopping event stream for %s.", serialnumber)
+            device.stop()
```
are these methods async? (because this is a sync context as the shutdown function has not been annotated)

Kane610 replied:
Do I need to decorate with coroutine in all sub functions for it to be a proper async context?

Kane610 replied:
Decorated with callback is that OK?

balloob commented on homeassistant/components/axis.py:
```diff
@@ -202,69 +209,71 @@ def vapix_service(call):
-                response = device.do_request(call.data[SERVICE_CGI],
-                                             call.data[SERVICE_ACTION],
-                                             call.data[SERVICE_PARAM])
+                response = device.vapix.do_request(call.data[SERVICE_CGI],
```
The service method is not annotated, so assuming this call is to a sync method ?

Kane610 replied:
At the moment it is a wrapper for the request library.
https://github.com/Kane610/axis/blob/7c143aebe2b03c3b7d2c2722dea947f27efe1dfa/axis/vapix.py#L40

balloob commented on homeassistant/components/axis.py:
```diff
@@ -202,69 +209,71 @@ def vapix_service(call):
-    device_config['hass'] = hass
-    device = AxisDevice(device_config)  # Initialize device
-    enable_metadatastream = False
+    def signal_callback(action, event):
```
Is this method run on the event loop by axis? If so, please annotate with `@callback`  (imported from `homeassistant.core`)

Kane610 replied:
The function is not decorated with coroutine, it is originating from a asyncio datagram endpoint https://github.com/Kane610/axis/blob/7c143aebe2b03c3b7d2c2722dea947f27efe1dfa/axis/rtsp.py#L168
and then finally called from here https://github.com/Kane610/axis/blob/7c143aebe2b03c3b7d2c2722dea947f27efe1dfa/axis/event.py#L85

Kane610 replied:
To my understanding, the datagram endpoint is callback based so I guess it is not run on the event loop

armills replied:
The datagram endpoint will be called on the event loop, so it's safe to annotate it as `@callback`.

balloob commented on homeassistant/components/axis.py:
```diff
@@ -287,27 +296,19 @@ def _write_config(hass, config):
+        self.location = event_config[ATTR_LOCATION]
         self.axis_event.callback = self._update_callback
 
+    @callback
```
You can't call `schedule_update_ha_state` from an async context.

Kane610 replied:
Ok, will look in to this.

I'm not fully getting the async context thing yet.

Kane610 replied:
As I understand it, the chain of methods are not on the event loop so I will remove the @callback

Kane610 committed:
Fix review comments

Kane610 commented:
Thinking a bit about this and seeing @balloob comments I think I need to go through everything a bit more and better understand asyncio before making this ready for integration. So I'm changing the state to a WIP.

Kane610 committed:
Removed async after some discussions with @armills.
It wasn't possible to get past that aiohttp doesnt support digest auth, which makes it impossible to convert the full library at this point in time.

Kane610 commented:
This is now just a small refactorization of the component and library with emphasis on removing the external gstreamer dependencies.
